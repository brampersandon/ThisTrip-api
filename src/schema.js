// @flow
import { gql } from 'apollo-server'

import { Provider, Route, Direction, Departure } from './models'

import type { Axios } from 'axios'
import type { CacheClient } from './cache'

export const typeDefs = gql`
  type Provider {
    name: String
    id: String
  }

  type Route {
    name: String
    providerId: String
    id: String
  }

  type Direction {
    id: String
    name: String
  }

  type Departure {
    departingIn: String
    routeId: String
    direction: String
    latitude: Float
    longitude: Float
  }

  type Stop {
    name: String
    description: String
    id: String
    latitude: Float
    longitude: Float
    distance: Float
  }

  type Query {
    routes: [Route]
    providers: [Provider]
    nearbyStops(
      latitude: Float
      longitude: Float
      radius: Float
      units: String
    ): [Stop]
    stop(id: String): Stop
    departures(stopId: String): [Departure]
    directions(routeId: String): [Direction]
  }
`

export const createResolvers = (cache: CacheClient, axios: Axios) => ({
  Query: {
    routes: async () => {
      const response = await axios.get(Route.url)
      if (response && response.data && response.data.length > 0)
        return response.data.map(Route.parse)
    },
    providers: async () => {
      const response = await axios.get(Provider.url)
      if (response && response.data && response.data.length > 0)
        return response.data.map(Provider.parse)
    },
    departures: async (_: *, { stopId }: { stopId: string }) => {
      const response = await axios.get(Departure.url(stopId))
      if (response && response.data && response.data.length > 0)
        return response.data.map(Departure.parse).filter(d => d != null)
    },
    nearbyStops: async (
      _: *,
      {
        latitude,
        longitude,
        radius = 500,
        units = 'm'
      }: { latitude: string, longitude: string, radius: number, units: string }
    ) => {
      const response = await cache.getStopsInRadius(
        latitude,
        longitude,
        radius,
        units
      )
      // Response should be an array, else we can't decorate our stops correctly.
      if (!response || !response.length) return
      const stops = Promise.all(
        response
          .map(async ({ key, latitude, longitude, distance }) => {
            const stopResponse = await cache.getStop(key)
            if (!stopResponse) return null
            const { name, description, id } = stopResponse
            return { name, description, id, latitude, longitude, distance }
          })
          .filter(stop => stop != null)
      )
      return stops
    },
    stop: async (_: *, { id }: { id: string }) => {
      const response = await cache.getStop(id)
      if (response) return response
    },
    directions: async (_: *, { routeId }: { routeId: string }) => {
      const response = await axios.get(Direction.url(routeId))
      return response.data.map(Direction.parse)
    }
  }
})
