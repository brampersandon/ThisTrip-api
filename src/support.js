export const departures = [
  {
    Actual: true,
    BlockNumber: 1140,
    DepartureText: '7 Min',
    DepartureTime: '/Date(1533931200000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 44.99882,
    VehicleLongitude: -93.29575
  },
  {
    Actual: true,
    BlockNumber: 1137,
    DepartureText: '9 Min',
    DepartureTime: '/Date(1533931320000-0500)/',
    Description: 'Chicago - 56St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'B',
    VehicleHeading: 0,
    VehicleLatitude: 45.00329,
    VehicleLongitude: -93.29567
  },
  {
    Actual: true,
    BlockNumber: 1430,
    DepartureText: '12 Min',
    DepartureTime: '/Date(1533931500000-0500)/',
    Description: 'Cedar-28Av/VA Med Ctr/Via Dupont-57',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'H',
    VehicleHeading: 0,
    VehicleLatitude: 45.01678,
    VehicleLongitude: -93.28809
  },
  {
    Actual: false,
    BlockNumber: 1146,
    DepartureText: '3:06',
    DepartureTime: '/Date(1533931560000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 44.98563,
    VehicleLongitude: -93.28528
  },
  {
    Actual: false,
    BlockNumber: 1131,
    DepartureText: '3:14',
    DepartureTime: '/Date(1533932040000-0500)/',
    Description: 'Chicago - 38St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'A',
    VehicleHeading: 0,
    VehicleLatitude: 45.03725,
    VehicleLongitude: -93.31191
  },
  {
    Actual: false,
    BlockNumber: 1428,
    DepartureText: '3:17',
    DepartureTime: '/Date(1533932220000-0500)/',
    Description: 'Cedar-28Av/VA Med Ctr/Via 51-Penn',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'H',
    VehicleHeading: 0,
    VehicleLatitude: 45.04391,
    VehicleLongitude: -93.30897
  },
  {
    Actual: false,
    BlockNumber: 1147,
    DepartureText: '3:21',
    DepartureTime: '/Date(1533932460000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 0.0,
    VehicleLongitude: 0.0
  },
  {
    Actual: false,
    BlockNumber: 1138,
    DepartureText: '3:29',
    DepartureTime: '/Date(1533932940000-0500)/',
    Description: 'Chicago - 38St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'A',
    VehicleHeading: 0,
    VehicleLatitude: 0.0,
    VehicleLongitude: 0.0
  },
  {
    Actual: false,
    BlockNumber: 1422,
    DepartureText: '3:32',
    DepartureTime: '/Date(1533933120000-0500)/',
    Description: 'Cedar-28Av/VA Med Ctr/Via Logan-57',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'G',
    VehicleHeading: 0,
    VehicleLatitude: 45.06057,
    VehicleLongitude: -93.31935
  },
  {
    Actual: false,
    BlockNumber: 1683,
    DepartureText: '3:36',
    DepartureTime: '/Date(1533933360000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 0.0,
    VehicleLongitude: 0.0
  },
  {
    Actual: false,
    BlockNumber: 1136,
    DepartureText: '3:45',
    DepartureTime: '/Date(1533933900000-0500)/',
    Description: 'Chicago - 38St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'A',
    VehicleHeading: 0,
    VehicleLatitude: 45.03548,
    VehicleLongitude: -93.29762
  },
  {
    Actual: false,
    BlockNumber: 1440,
    DepartureText: '3:48',
    DepartureTime: '/Date(1533934080000-0500)/',
    Description: 'Cedar-28Av/VA Med Ctr/Via Dupont',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'G',
    VehicleHeading: 0,
    VehicleLatitude: 0.0,
    VehicleLongitude: 0.0
  },
  {
    Actual: false,
    BlockNumber: 1135,
    DepartureText: '3:53',
    DepartureTime: '/Date(1533934380000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 44.97332,
    VehicleLongitude: -93.26372
  },
  {
    Actual: false,
    BlockNumber: 1149,
    DepartureText: '4:00',
    DepartureTime: '/Date(1533934800000-0500)/',
    Description: 'Chicago - 38St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'A',
    VehicleHeading: 0,
    VehicleLatitude: 0.0,
    VehicleLongitude: 0.0
  },
  {
    Actual: false,
    BlockNumber: 1434,
    DepartureText: '4:03',
    DepartureTime: '/Date(1533934980000-0500)/',
    Description: 'Cedar-28Av/VA Med Ctr/Via 51-Penn',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'H',
    VehicleHeading: 0,
    VehicleLatitude: 0.0,
    VehicleLongitude: 0.0
  },
  {
    Actual: false,
    BlockNumber: 1142,
    DepartureText: '4:07',
    DepartureTime: '/Date(1533935220000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 45.0059,
    VehicleLongitude: -93.29413
  },
  {
    Actual: false,
    BlockNumber: 1139,
    DepartureText: '4:16',
    DepartureTime: '/Date(1533935760000-0500)/',
    Description: 'Chicago - 38St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'A',
    VehicleHeading: 0,
    VehicleLatitude: 44.97891,
    VehicleLongitude: -93.27707
  },
  {
    Actual: false,
    BlockNumber: 1431,
    DepartureText: '4:18',
    DepartureTime: '/Date(1533935880000-0500)/',
    Description: 'Cedar-28Av/VA Med Ctr/Via Logan-57',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'G',
    VehicleHeading: 0,
    VehicleLatitude: 44.97669,
    VehicleLongitude: -93.27161
  },
  {
    Actual: false,
    BlockNumber: 1132,
    DepartureText: '4:23',
    DepartureTime: '/Date(1533936180000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 44.87566,
    VehicleLongitude: -93.26798
  },
  {
    Actual: false,
    BlockNumber: 1144,
    DepartureText: '4:31',
    DepartureTime: '/Date(1533936660000-0500)/',
    Description: 'Chicago - 38St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'A',
    VehicleHeading: 0,
    VehicleLatitude: 44.96254,
    VehicleLongitude: -93.2625
  },
  {
    Actual: false,
    BlockNumber: 1436,
    DepartureText: '4:32',
    DepartureTime: '/Date(1533936720000-0500)/',
    Description: 'Cedar-28Av/VA Med Ctr/Via Dupont-57',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'H',
    VehicleHeading: 0,
    VehicleLatitude: 44.95184,
    VehicleLongitude: -93.2474
  },
  {
    Actual: false,
    BlockNumber: 1112,
    DepartureText: '4:45',
    DepartureTime: '/Date(1533937500000-0500)/',
    Description: 'Chicago - 56St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'B',
    VehicleHeading: 0,
    VehicleLatitude: 44.94266,
    VehicleLongitude: -93.26264
  },
  {
    Actual: false,
    BlockNumber: 1429,
    DepartureText: '4:47',
    DepartureTime: '/Date(1533937620000-0500)/',
    Description: 'Cedar-28Av / VA Med Ctr/Via 51-Penn',
    Gate: '',
    Route: '22',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'G',
    VehicleHeading: 0,
    VehicleLatitude: 44.9346,
    VehicleLongitude: -93.2296
  },
  {
    Actual: false,
    BlockNumber: 1134,
    DepartureText: '4:52',
    DepartureTime: '/Date(1533937920000-0500)/',
    Description: 'Chicago / Mall of Amer',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'E',
    VehicleHeading: 0,
    VehicleLatitude: 44.89764,
    VehicleLongitude: -93.236
  },
  {
    Actual: false,
    BlockNumber: 1133,
    DepartureText: '5:00',
    DepartureTime: '/Date(1533938400000-0500)/',
    Description: 'Chicago - 56St',
    Gate: '',
    Route: '5',
    RouteDirection: 'SOUTHBOUND',
    Terminal: 'B',
    VehicleHeading: 0,
    VehicleLatitude: 44.93391,
    VehicleLongitude: -93.26295
  }
]

export const stops = [
  {
    name: '2nd St NE - Node & Broadway St',
    id: '12252'
  },
  {
    name: 'Broadway St NE & Marshall St NE',
    id: '12256'
  },
  {
    name: 'Broadway St NE & 2nd St NE',
    id: '12258'
  }
]
