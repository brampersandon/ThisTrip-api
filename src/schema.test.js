import { typeDefs, createResolvers } from './schema'

import { departures, stops } from './support'

const createMockRedis = (stopsData, stopData) => ({
  getStopsInRadius: async (..._) => Promise.resolve(stopsData),
  getStop: async _key => Promise.resolve(stopData)
})
const createMockAxios = data => ({ get: _url => Promise.resolve({ data }) })

// Resolver spec structure inspired in part by https://hackernoon.com/extensive-graphql-testing-57e8760f1c25
describe('resolvers', () => {
  const stopId = '12345'
  describe('departures', () => {
    it('can handle success', async () => {
      const redis = () => null
      const axios = createMockAxios(departures)
      const resolvers = createResolvers(redis, axios)

      const result = await resolvers.Query.departures({}, { stopId })
      expect(result).toMatchSnapshot()
    })
    it('can handle failure', async () => {
      const redis = () => null
      // Mock a bad HTTP response
      const axios = createMockAxios(false)
      const resolvers = createResolvers(redis, axios)

      const result = await resolvers.Query.departures({}, { stopId })

      // Expect the query to fail.
      expect(result).toBe(undefined)
    })
  })
  describe('stop', () => {
    it('can handle success', async () => {
      const redis = createMockRedis(stops, stops[0])
      const axios = () => null
      const resolvers = createResolvers(redis, axios)

      const result = await resolvers.Query.stop({}, { id: stopId })
      expect(result).toMatchSnapshot()
    })
    it('can handle failure', async () => {
      const redis = createMockRedis(false)
      const axios = () => null
      const resolvers = createResolvers(redis, axios)

      const result = await resolvers.Query.stop({}, { id: stopId })

      // Expect the query to fail.
      expect(result).toBe(undefined)
    })
  })
})
describe('nearbyStops', () => {
  const coords = { latitude: 45.0, longitude: -93.0 }
  it('can handle success', async () => {
    const redis = createMockRedis(stops, stops[0])
    const axios = () => null
    const resolvers = createResolvers(redis, axios)

    const result = await resolvers.Query.nearbyStops({}, coords)
    expect(result).toMatchSnapshot()
  })
  it('can handle failure', async () => {
    const redis = createMockRedis(false, false)
    const axios = () => null
    const resolvers = createResolvers(redis, axios)

    const result = await resolvers.Query.nearbyStops({}, coords)

    // Expect the query to fail.
    expect(result).toBe(undefined)
  })
})
xdescribe('routes')
xdescribe('providers')
xdescribe('directions')
