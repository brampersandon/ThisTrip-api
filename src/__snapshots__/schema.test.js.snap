// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nearbyStops can handle success 1`] = `
Array [
  Object {
    "description": undefined,
    "distance": undefined,
    "id": "12252",
    "latitude": undefined,
    "longitude": undefined,
    "name": "2nd St NE - Node & Broadway St",
  },
  Object {
    "description": undefined,
    "distance": undefined,
    "id": "12252",
    "latitude": undefined,
    "longitude": undefined,
    "name": "2nd St NE - Node & Broadway St",
  },
  Object {
    "description": undefined,
    "distance": undefined,
    "id": "12252",
    "latitude": undefined,
    "longitude": undefined,
    "name": "2nd St NE - Node & Broadway St",
  },
]
`;

exports[`resolvers departures can handle success 1`] = `
Array [
  Object {
    "departingIn": "7 Min",
    "direction": "SOUTHBOUND",
    "latitude": 44.99882,
    "longitude": -93.29575,
    "routeId": "5",
  },
  Object {
    "departingIn": "9 Min",
    "direction": "SOUTHBOUND",
    "latitude": 45.00329,
    "longitude": -93.29567,
    "routeId": "5",
  },
  Object {
    "departingIn": "12 Min",
    "direction": "SOUTHBOUND",
    "latitude": 45.01678,
    "longitude": -93.28809,
    "routeId": "22",
  },
]
`;

exports[`resolvers stop can handle success 1`] = `
Object {
  "id": "12252",
  "name": "2nd St NE - Node & Broadway St",
}
`;
